torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,160 @@ namespace rpc {
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+
+    // Updating these since something may have been added to the map while this
+    // thread was sleeping.
+    earliestTimeout = rpcRetryMap_.begin()->first;
+    auto& earliestRpcList = rpcRetryMap_.begin()->second;
+
+    // We iterate through all the RPC's set to be retried at the current
+    // timepoint, resend those RPC's, and add a callback that either schedules
+    // the RPC for a future retry or marks it with success/error depending on
+    // the outcome of the current send. Then, we clean up the rpcRetryMap_.
+    for (auto it = earliestRpcList.begin(); it != earliestRpcList.end();
+         /* no increment */) {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+      earliestRpc->retryCount_++;
+
+      fm->addCallback([this, newTime, earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        rpcRetryCallback(message, futErr, newTime, earliestRpc);
+      });
+
+      // We have either scheduled the RPC for a future retry, or we've marked
+      // it complete upon success, or we've marked it with an error after
+      // maxRetries unsuccessful send attempts. Thus, we can remove it from the
+      // retry Map for the current timepoint.
+      it = earliestRpcList.erase(it);
+    }
+
+    // If there are no more RPC's set to be retried at the current timepoint,
+    // we can remove the corresponsing unordered_set from the retry map.
+    if (earliestRpcList.empty()) {
+      rpcRetryMap_.erase(earliestTimeout);
+    }
+  }
+}
+
+void RpcAgent::rpcRetryCallback(
+    const rpc::Message& message,
+    const c10::optional<utils::FutureError>& futErr,
+    steady_clock_time_point newTime,
+    std::shared_ptr<RpcRetryInfo> earliestRpc) {
+  if (futErr) {
+    LOG(INFO) << "Send try " << std::to_string(earliestRpc->retryCount_)
+              << " failed";
+    if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
+      // If the previous future completed with an error and we haven't
+      // completed maxRetries send attempts, we move the earliestRpc
+      // struct to a new time point in the retry map (effectively
+      // scheduling it for a future retry.)
+      int retryCount = earliestRpc->retryCount_;
+      {
+        std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+        rpcRetryMap_[newTime].emplace(std::move(earliestRpc));
+      }
+      if (retryCount == 1) {
+        // The retry thread waits for the map to be populated. Thus we notify
+        // once an item has been added the first time.
+        rpcRetryMapCV_.notify_one();
```

Comment:
Why do we have this specific check? Why can't we just notify always after we emplace in rpcRetryMap_ above?